### PR TITLE
[improve][meta] Support set metadata size threshold for compression.

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
[PIP-270](https://github.com/apache/pulsar/issues/20307)

### Motivation

Current we support to set compression on ManagedLedger and ManagedCursor metadata
by managedLedgerInfoCompressionType and managedCursorInfoCompressionType

the metadata of ManagedLedger is meanly composed by a ledger id list and some other field
to describe pulsar topic storage state. you can use the metadata to find which the message is stored.
when pulsar topic message is deleled by the retention policy or ledger rollover. the metadata is remove or append the metadata list.

this metadata can be small when retention time is short.

the metadata of ManagedCursor is the subscription state of a pulsar subscription. if user ack an message this will trigger the subscription state save in the metadata of ManagedCursor

if user always use accumlateAcknowledge api this state is also small enough.

all those metadata is saved in metadataStore.

sometimes the metadata is too small to compress. so a size based config is needed.

### Modifications

check size before metadata compression.

### Verifying this change
This change added tests and can be verified as follows:

add unit test case for metadata big or small than threshold.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Comment by lifepuzzlefun** [comment]
@codelipenghui @Technoboy- @AnonHxy plesase take a look. thank you.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java [hunk 1]
```diff
@@ -92,8 +92,30 @@ public class ManagedLedgerFactoryConfig {
...
      */
     private String managedCursorInfoCompressionType = MLDataFormats.CompressionType.NONE.name();
+
+    /**
+     * ManagedCursorInfo compression threshold. If the origin metadata size below configuration.
+     * compression will not apply.
+     */
+    private int managedCursorInfoCompressionSizeThresholdBytes = 0;
+
+    public MetadataCompressionConfig getCompressionConfigForManagedLedgerInfo() {
```
Comment (AnonHxy) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java:112]: There is no need to add getter  method here

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java [hunk 2]
```diff
@@ -92,8 +92,30 @@ public class ManagedLedgerFactoryConfig {
      */
     private String managedLedgerInfoCompressionType = MLDataFormats.CompressionType.NONE.name();
 
+    /**
+     * ManagedLedgerInfo compression threshold. If the origin metadata size below configuration.
+     * compression will not apply.
+     */
+    private int managedLedgerInfoCompressionSizeThresholdBytes = 0;
```
Comment (315157973) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java:99]: Should we set a suitable default value?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/MetaStoreImpl.java [hunk 3]
```diff
@@ -415,29 +400,43 @@ private static MetaStoreException getException(Throwable t) {
...
+            MLDataFormats.ManagedCursorInfoMetadata metadata = MLDataFormats.ManagedCursorInfoMetadata
+                    .newBuilder()
+                    .setCompressionType(compressionType)
+                    .setUncompressedSize(uncompressedSize)
+                    .build();
+            return compressManagedInfo(managedCursorInfo.toByteArray(), metadata.toByteArray(),
+                    metadata.getSerializedSize(), compressionType);
+        }
+
+        return managedCursorInfo.toByteArray();
```
Comment (315157973) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/MetaStoreImpl.java:439]: What if it is uncompressed data when decompressing?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java (hunk 1 above)
Comment (lifepuzzlefun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java:112]: It is for create MetadataCompressionConfig which encapsulate the compression type and threshold。

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java (hunk 2 above)
Comment (lifepuzzlefun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactoryConfig.java:99]: current value is the same behavior before this feature. maybe we can set to 128 or more. In our production env most cursor metadata is less than this value.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/MetaStoreImpl.java (hunk 3 above)
Comment (lifepuzzlefun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/MetaStoreImpl.java:439]: the compressed data is self-explain with `MLDataFormats.ManagedLedgerInfoMetadata` padding before the compressed data content. the parse logic will try to read magic number bytes. if magic number not exists, the data will be thought no compression is applied.

**Review by 315157973 — APPROVED** [review]

**Comment by lifepuzzlefun** [comment]
@Technoboy- @congbobo184 @poorbarcode can you take a look ？

**Review by poorbarcode — APPROVED** [review]
LGTM. Left some comments

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/MetadataCompressionConfig.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        this(compressionType, 0);
+    }
+
+    public MetadataCompressionConfig(String compressionType, int compressThreshold) throws IllegalArgumentException {
+        this.compressionType = parseCompressionType(compressionType);
+        this.compressThreshold = compressThreshold;
+    }
+
+    public static MetadataCompressionConfig noCompression =
+            new MetadataCompressionConfig(MLDataFormats.CompressionType.NONE, 0);
```
[the remark this answers is not in the document]
Comment (lifepuzzlefun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/MetadataCompressionConfig.java:44]: done

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/MetadataCompressionConfig.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import lombok.ToString;
+import org.apache.bookkeeper.mledger.proto.MLDataFormats;
+import org.apache.commons.lang.StringUtils;
+
+@Data
+@AllArgsConstructor
+@ToString
+public class MetadataCompressionConfig {
+    MLDataFormats.CompressionType compressionType;
+    int compressThreshold;
```
[the remark this answers is not in the document]
Comment (lifepuzzlefun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/MetadataCompressionConfig.java]: done

**Comment by lifepuzzlefun** [comment]
@poorbarcode @Technoboy- @315157973 can this pr be merged. if need change please let me know : - )

**Comment by 315157973** [comment]
Some unit tests fail, please rebase  master

**Review by Technoboy- — APPROVED** [review]

**Comment by Technoboy-** [comment]
> @Technoboy- @congbobo184 @poorbarcode can you take a look ？

Sorry for the late response.

**Comment by 315157973** [comment]
@lifepuzzlefun  Please fix the unit test

**Comment by lifepuzzlefun** [comment]
@315157973 @poorbarcode @Technoboy- can this pr be merged？

**Comment by 315157973** [comment]
> @315157973 @poorbarcode @Technoboy- can this pr be merged？

The button is unavailable due to unit tests not all passing

**Review by congbobo184 — CHANGES_REQUESTED** [review]
Where can we change these config values? I haven't seen any opreation that we can change these config values. Maybe we should create a PIP to add these two configurations?

**Comment by lifepuzzlefun** [comment]
@congbobo184 aha, i forget add those config in the pulsar configuration. i have add them in the new commit : - )
and do we need a pip to merge this patch ? this is a small feature introduce to pulsar. may be i can update the doc to tell people the support settings.

**Comment by congbobo184** [comment]
@lifepuzzlefun hi, add any config, we also should create a PIP to discuss and vote

**Comment by lifepuzzlefun** [comment]
@congbobo184 hi, sorry for the late reply, i have just create a pip to discuss.

**Comment by lifepuzzlefun** [comment]
https://github.com/apache/pulsar/issues/20307 pip-270 created. hopes for your comments and vote, thank you~ 

the discuss mail list address is write at the bottom, feel free to join the discuss.

**Comment by Technoboy-** [comment]
@congbobo184 Could you take a review?

**Pull request merged**